subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/DefaultDependencyCoordinateResolverFactory.java

```diff
@@ -41,33 +46,79 @@
+
+            boolean useResolvedCoordinates = resolutionBackedVariant.getPublishResolvedCoordinates();
+            if (useResolvedCoordinates && configuration == null) {
+                throw new InvalidUserDataException("Cannot enable dependency mapping without configuring a resolution configuration.");
+            } else if (useResolvedCoordinates) {
+
+                ComponentDependencyResolver componentResolver = new ResolutionBackedComponentDependencyResolver(
+                    configuration,
+                    moduleIdentifierFactory,
+                    projectDependencyResolver
+                );
+
+                VariantDependencyResolver variantResolver = new ResolutionBackedVariantDependencyResolver(
+                    projectDependencyResolver,
+                    moduleIdentifierFactory,
+                    configuration,
+                    attributeDesugaring,
+                    componentResolver
+                );
+
+                return new DependencyResolvers(variantResolver, componentResolver);
+            }
+        }
 
         ImmutableAttributes attributes = ((AttributeContainerInternal) variant.getAttributes()).asImmutable();
         VariantVersionMappingStrategyInternal versionMapping = versionMappingStrategy.findStrategyForVariant(attributes);
 
         // Fallback to component coordinate mapping if variant mapping is not enabled
-        ComponentDependencyResolver componentResolver;
+        ComponentDependencyResolver componentResolver = null;
         if (versionMapping.isEnabled()) {
             if (versionMapping.getUserResolutionConfiguration() != null) {
                 configuration = versionMapping.getUserResolutionConfiguration();
-            } else if (versionMapping.getDefaultResolutionConfiguration() != null) {
+            } else if (versionMapping.getDefaultResolutionConfiguration() != null && configuration == null) {
+                // The configuration set on the variant is almost always more correct than the
+                // default version mapping configuration, which is currently set project-wide
+                // by the Java plugin. For this reason, we only use the version mapping default
+                // if the dependency mapping configuration is not set.
                 configuration = versionMapping.getDefaultResolutionConfiguration();
             }
+
+            if (configuration != null) {
+                componentResolver = USE_LEGACY_VERSION_MAPPING
+                    ? new VersionMappingVariantDependencyResolver(projectDependencyResolver, configuration)
+                    : new ResolutionBackedComponentDependencyResolver(configuration, moduleIdentifierFactory, projectDependencyResolver);
+            }
         }
 
-        if (configuration != null) {
-            componentResolver = new VersionMappingVariantDependencyResolver(projectDependencyResolver, configuration);
-        } else {
+        if (componentResolver == null) {
```

Comment:
💅 Better as just an else on the `isEnabled()` if.

Reply:
It is possible for the `versionMapping.isEnabled()` branch to exit with the `componentResolver` initialized -- specifically when the user has not configured a version mapping configuration and when there is no default version mapping configuration set by the version mapping strategy

Reply:
Oh, OK, I see the case now.